### ISSUE TITLE
[no-inferrable-types] Add an option to enable it for any values

### ISSUE THREAD

**Issue by ark120202** [typescript-eslint/typescript-eslint#907]
Currently this rule warns only when rhs is either a literal or an associated global constructor function. I'd like to have an option to warn about any cases when type annotation doesn't change variable type.

```ts
type MyType = { foo: string };
const getString = () => 'foo';

const valid1: string | undefined = getString();
const valid2: MyType = { foo: 'bar' };

const invalid1: string = getString();
const invalid2: string | string = getString();
const invalid3: string | undefined = Math.random() > 0.5 ? getString() : undefined;
class Test { static invalid4: Test = new Test(); }
const invalid5: MyType = null! as MyType;
```

I've done a bit of testing and apparently equal types have the same reference (even in the `invalid2` case), so something like `typeChecker.getTypeAtLocation(typeAnnotation) === typeChecker.getTypeAtLocation(initializer)` can be enough to check this.

**Comment by ark120202** [comment]
No, none of the cases listed there are handled by `no-unnecessary-type-assertion`.

Wouldn't it be possible to make it require type checker only when a new option is enabled?

**Comment by ark120202** [comment]
> I guess the naming's kinda bad, but the idea is that you can easily infer the type from a primitive value, even without an IDE, so strictly typing it is literally useless.

Makes sense, would it worth it to rename it then? Maybe to something like `obvious` or `apparent`? Also [TypeScript documentation link](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-inferrable-types.md#further-reading) doesn't seem really related in that case.

> So I'd probably lean toward this belong in `nuta`, because strictly typing a variable/property/argument declaration is a type assertion on the value.

[Type assertion](https://www.typescriptlang.org/docs/handbook/basic-types.html#type-assertions) has a strict definition in typescript docs, so I think that may make it a bit confusing. Maybe it would make sense to rename it to `no-unnecessary-types` that would handle both type assertions and type annotations?

**Comment by glen-84** [comment]
Is this a duplicate of #295?

**Comment by bradzacher** [comment]
actually, I think it is - good catch.
